### ISSUE TITLE
convert and run opencv example ok, but classification result is differnt

### ISSUE THREAD

**Issue by glocktsai** [Dobiasd/frugally-deep#108]
[52_final.tar.gz](https://github.com/Dobiasd/frugally-deep/files/2694202/52_final.tar.gz)
Hi Tobias Hermann,
    thank you for your contribution of frugally-deep. it's a great help to deploy keras model.
    however the classification result is different.
    would you please help to check attached file?
    thank you very much.

**Comment by Dobiasd** [comment]
Hi glocktsai,

can you please provide a minimal C++ example code to reproduce the issue?

**Comment by glocktsai** [comment]
sorry I update command line information as below, and update zip file again.
[52_final.tar.gz](https://github.com/Dobiasd/frugally-deep/files/2694324/52_final.tar.gz)

keras inference: python3 52cat.py
C++ inference: opencv_example 5_card_1.jpg

**Comment by Dobiasd** [comment]
Your model (`52cat.h5`) takes input images of size 224x224.

Your image (`5_card_1.jpg`) has size 67x40.

In Python you upscale the image to 224x224. In C++ you currently don't.

Please resize the image externally (with Gimp or something) to 224x224 and save it again (ideally as `.png` to avoid quality loss).

Then remove the resizing from your Python code and just load the new image as is.
This way you can make sure that in C++ you really give the models the same input values as you do in Python. If then the result is correct, i.e., same prediction results in C++ as in Python, you can try to add the resizing to your C++ code too. But first try to make sure the model itself works as intended, and avoid code around it that might taint this measurement.

**Comment by glocktsai** [comment]
I change to the test image to 224x224 in png format. and load this file without scaling in python and c++.
python3 52cat.py => 224.png:9:10_spades_10 => class 9
opencv_example 224.png => class 36
the updated file are attached, would you please try it?
thank you very much.
[224png.tar.gz](https://github.com/Dobiasd/frugally-deep/files/2697938/224png.tar.gz)

**Comment by Dobiasd** [comment]
Usually the pixel values of images that go as input to a model are scaled into the interval `[0.0, 1.0)`. In the example you show with your model however this not the case. The pixel values are left in `[0, 256)`. In that case you have to do the same on C++ side.

Here is a minimal example in Python and in C++, that both output the same prediction with your model and image (224x224), i.e., the class `9`:

```python
import keras
from keras.preprocessing import image
from keras.models import load_model
import numpy as np

model = load_model('52cat.h5')

img = image.load_img("224.png")
img_array = image.img_to_array(img)
img_array_expanded_dims = np.expand_dims(img_array, axis=0)

prediction = model.predict(img_array_expanded_dims).argmax(axis=-1)[0]
print(prediction)
```

```C++
// g++ -std=c++14 -O3 opencv_example.cpp -lopencv_core -lopencv_imgproc -lopencv_imgcodecs -o opencv_example

#include <fdeep/fdeep.hpp>
#include <opencv2/opencv.hpp>

int main(int argc, char **argv)
{
    const cv::Mat image = cv::imread("224.png");
    cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
    assert(image.isContinuous());
    const auto model = fdeep::load_model("52cat.json");
    const auto input = fdeep::tensor5_from_bytes(image.ptr(),
        static_cast<std::size_t>(image.rows),
        static_cast<std::size_t>(image.cols),
        static_cast<std::size_t>(image.channels()),
        0.0f, 255.0f); // Note the scaling here!
    const auto result = model.predict_class({input});
    std::cout << result << std::endl;
    return result;
}
```

Please also note that I removed all the unnecessary fluff code from your Python example. It just makes it harder to track the actual problem. Always try to reduce your examples to the bare minimum needed. This not only makes it easier for others to understand your issue, but also helps yourself in finding the cause on your own.

The next debugging step would have been to print out the actual values going into the model in C++ and in Python and compare them. In that case you would also quickly have realized that the difference in model output is simply caused by different input.

**Comment by glocktsai** [comment]
thank you very much, i forgot the pre-processing.
I checked keras source code, the default is scaling to [-1.0, +1.0]  when backend is tensorflow
so i modified to 
    const auto input = fdeep::tensor5_from_bytes(image.ptr(),
        static_cast<std::size_t>(image.rows),
        static_cast<std::size_t>(image.cols),
        static_cast<std::size_t>(image.channels()),
        -1.0f, 1.0f);
and it works now!!
I will check remaining pictures later.
thank you very much.

**Comment by Dobiasd** [comment]
Oh, so actually it's `[-1.0, 1.0)`. Thanks. It's interesting that the model also seems to somewhat work correctly with `[0, 256)`, at least on this one example image.

**Comment by Dobiasd** [comment]
Thanks for the extended analysis and the feedback. The performance results are surprisingly good, but I'll not complain about this. :grin: :racing_car:
